src/main/java/blusunrize/immersiveengineering/common/items/SurveyToolsItem.java

```diff
@@ -0,0 +1,230 @@
+		});
+		// soft rocks
+		CAN_USE_ON.add((world, pos) -> {
+			BlockState state = world.getBlockState(pos);
+			return state.getMaterial()==Material.ROCK&&state.getBlockHardness(world, pos) < 0.5;
+		});
+	}
+
+	@Override
+	public void addInformation(ItemStack stack, @Nullable World worldIn, List<ITextComponent> tooltip, ITooltipFlag flagIn)
+	{
+		super.addInformation(stack, worldIn, tooltip, flagIn);
+	}
+
+	@Override
+	public UseAction getUseAction(ItemStack stack)
+	{
+		return UseAction.BLOCK;
+	}
+
+	@Override
+	public int getUseDuration(ItemStack stack)
+	{
+		return 50;
+	}
+
+	@Override
+	public ActionResultType onItemUseFirst(ItemStack stack, ItemUseContext context)
+	{
+		PlayerEntity player = context.getPlayer();
+		if(player==null)
+			return ActionResultType.PASS;
+		World world = context.getWorld();
+		BlockPos pos = context.getPos();
+		if(!CAN_USE_ON.stream().anyMatch(predicate -> predicate.test(world, pos)))
+		{
+			player.sendStatusMessage(new TranslationTextComponent(Lib.CHAT_INFO+"survey.wrong_block"), true);
+			return ActionResultType.FAIL;
+		}
+		player.setActiveHand(context.getHand());
+		return ActionResultType.SUCCESS;
+	}
+
+	@Override
+	public ItemStack onItemUseFinish(ItemStack stack, World world, LivingEntity entityLiving)
+	{
+		if(!(entityLiving instanceof ServerPlayerEntity))
+			return stack;
+		ServerPlayerEntity player = (ServerPlayerEntity)entityLiving;
+		RayTraceResult rtr = rayTrace(world, player, FluidMode.NONE);
+		if(!(rtr instanceof BlockRayTraceResult))
+			return stack;
+		BlockPos pos = ((BlockRayTraceResult)rtr).getPos();
+		MineralVein vein = ExcavatorHandler.getRandomMineral(world, pos);
+		if(vein==null)
+		{
+			player.sendStatusMessage(new TranslationTextComponent(Lib.CHAT_INFO+"survey.no_vein"), true);
+			return stack;
+		}
+
+		ListNBT data = getVeinData(stack, world.getDimension().getType().getRegistryName().toString(), vein.getPos());
+		int dataCount = data.size();
+		/* I considered not giving any information after 3 surveys, but because the text is displayed above the action
+		 * bar and can't be brought back after it fades, that could lead to frustration
+
+		if(dataCount >= 3)
+		{
+			player.sendStatusMessage(new TranslationTextComponent(Lib.CHAT_INFO+"survey.finished"), true);
+			return stack;
+		}
+		 */
+
+		// Need at least 4 blocks between samples
+		boolean tooClose = data.stream().anyMatch(inbt -> {
+			int dX = ((CompoundNBT)inbt).getInt("x")-pos.getX();
+			int dZ = ((CompoundNBT)inbt).getInt("z")-pos.getZ();
+			return dX*dX+dZ*dZ < 16;
+		});
+		if(tooClose)
+		{
+			player.sendStatusMessage(new TranslationTextComponent(Lib.CHAT_INFO+"survey.too_close"), true);
+			return stack;
+		}
+
+		// Get angle between postion->center vector and standard (south facing) vector
+		Vector2f vecToCenter = new Vector2f(vein.getPos().x-pos.getX(), vein.getPos().z-pos.getZ());
+		double angle = Math.toDegrees(Math.atan2(vecToCenter.y, vecToCenter.x));
```

Comment:
What does this do when you sample at the vein center?

Reply:
It puts out "0 steps to the east". Could probably specialcase it.

Reply:
Would probably be a good idea from a UI perspective. I was a more worried about having `NaN`s/`Infinity`s running around there, but it looks like `atan2` just returns 0 for `(0, 0)`.